### PR TITLE
Enable scale up/down

### PR DESCRIPTION
**Pull request by kargakis**
Ref https://jira.coreos.com/browse/AZURE-102
Fixes https://github.com/openshift/openshift-azure/issues/285

@openshift/sig-azure

### CODE REVIEW

File: cmd/createorupdate/update.go [hunk 1]
```diff
@@ -43,6 +48,42 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
...
+	for _, agent := range oldCs.Properties.AgentPoolProfiles {
+		if counts[agent.Name] != agent.Count {
+			bm["scale"] = true
+		}
+	}
+
+	return bm
+}
+
+func waitForReady(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin, vmc compute.VirtualMachineScaleSetVMsClient) error {
```
Comment (kargakis) [inline cmd/createorupdate/update.go]: Should probably cross-check with scale set count

File: cmd/createorupdate/update.go [hunk 2]
```diff
@@ -43,6 +52,42 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
 	return nil
 }
 
+// needUpdate returns a map with the exact operations that need to be
+// executed during an upgrade. Only scale is supported for now.
+func needUpdate(cs, oldCs *api.OpenShiftManagedCluster) map[string]bool {
```
Comment (kargakis) [inline cmd/createorupdate/update.go]: Eventually we will need to pull in here azuredeploy.json (both old and new) in order to figure out whether the CSE has changed.

File: pkg/validate/validate.go [hunk 3]
```diff
@@ -68,8 +65,21 @@ func validateContainerService(c *api.OpenShiftManagedCluster, externalOnly bool)
 }
 
 func validateUpdateContainerService(cs, oldCs *api.OpenShiftManagedCluster, externalOnly bool) (errs []error) {
-	if !reflect.DeepEqual(cs, oldCs) {
```
Comment (kargakis) [inline pkg/validate/validate.go]: @jim-minter not sure why these changes were reverted in https://github.com/openshift/openshift-azure/pull/238 - https://github.com/openshift/openshift-azure/pull/181 was a more reasonable approach to validate updates vs invalidating everything.

**Comment by kargakis** [comment]
We need to identify requests for scale down before we apply the ARM template, in order to run drain.

File: cmd/manifest/manifest.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
+package main
```
Comment (jim-minter) [inline cmd/manifest/manifest.go]: This is grim.  Do we really need to solve this problem here?  End users can just edit their manifest yaml.  Scale testing can create its own manifest yaml and modify it using its own separate codebase

File: cmd/createorupdate/update.go [hunk 5]
```diff
@@ -23,6 +23,15 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
 	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
 	vmc.Authorizer = authorizer
 
+	bm := needUpdate(cs, oldCs)
+	if len(bm) == 1 && bm["scale"] {
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: I don't think this is right yet because needUpdate is only partially implemented.  It should fire only if nothing but counts (or a single count?) has changed.  Currently it would fire if a count + 0 or more additional things had changed.

File: cmd/createorupdate/update.go [hunk 6]
```diff
@@ -23,6 +23,15 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
...
+		// At this point the scale set has already been updated so we
+		// only need to wait for things (VMs) to settle.
+		// TODO: Figure out whether scale down is going to be problematic
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: I think if we're doing a scale hack, it should include this too.

File: cmd/createorupdate/update.go (hunk 2 above)
Comment (jim-minter) [inline cmd/createorupdate/update.go]: I can't say with any certainty that I think we'll actually do this.  For now I think this PR needs to run deploy() within update() instead of outside, and it needs a hack before that which detects if the only changes are scale-related and actions them.

File: cmd/createorupdate/update.go (hunk 1 above)
Comment (jim-minter) [reply cmd/createorupdate/update.go]: if you have a custom hack which actions scales, it should be able to detect new vms created and just call WaitForReady() on them.  I don't think we shouldn't call it on all of them in case one is broken.
similarly it should have the code to drain vms that will be deleted.  So I think this will need to be refactored a little.

Comment (kargakis) [reply cmd/createorupdate/update.go]: >  similarly it should have the code to drain vms that will be deleted. So I think this will need to be refactored a little.

In case of scale down, drain needs to run before we apply the ARM template, as we discussed yesterday, right?

File: cmd/manifest/manifest.go (hunk 4 above)
Comment (kargakis) [reply cmd/manifest/manifest.go]: I don't care where this lives but it's something that can evolve to a e2e library.

File: pkg/validate/validate.go (hunk 3 above)
Comment (kargakis) [reply pkg/validate/validate.go]: There are a couple of extra fields that we need to gate apart from the current, `vmSize`, `publicHostname`, and `osType`. I can invalidate changes to those too, error messages are going to be friendlier.

File: cmd/createorupdate/update.go (hunk 5 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Right, this needs more work

File: cmd/createorupdate/update.go (hunk 1 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: After moving `deploy` in `update` this made sense.

File: pkg/validate/validate.go (hunk 3 above)
Comment (kargakis) [reply pkg/validate/validate.go]: Opted for your suggestion as it was much less code, better for review and the error message is ok.

File: cmd/createorupdate/update.go (hunk 2 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Right, removed in favor of a special cased function for scaling.

File: cmd/manifest/manifest.go (hunk 4 above)
Comment (kargakis) [reply cmd/manifest/manifest.go]: removed for now, didnt add much indeed

**Comment by kargakis** [comment]
@jim-minter updated as per your suggestions and added a unit test, ptal

File: cmd/createorupdate/update.go [hunk 7]
```diff
@@ -11,7 +12,19 @@ import (
 	"github.com/openshift/openshift-azure/pkg/log"
 )
 
-func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin) error {
+func update(ctx context.Context, cs, oldCs *api.OpenShiftManagedCluster, p api.Plugin, azuredeploy []byte) error {
+	scaleOps := isScaleOp(cs, oldCs)
+	for _, scale := range scaleOps {
+		if scale.down {
+			// TODO: If we are scaling down, we need to drain first.
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: We probably want to do this always, regardless of whether it's a scale op or a more complex update.
The kernel of the code is at https://github.com/openshift/openshift-azure/blob/master/cmd/createorupdate/update.go#L131 - just factor it out, loop the roles here and call it.

File: cmd/createorupdate/update.go [hunk 8]
```diff
@@ -11,7 +12,19 @@ import (
...
+		}
+	}
+
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: Here, loop roles and for each, record which VMs currently exist pre-deploy

File: cmd/createorupdate/update.go [hunk 9]
```diff
@@ -23,6 +36,19 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
 	vmc := compute.NewVirtualMachineScaleSetVMsClient(cs.Properties.AzProfile.SubscriptionID)
 	vmc.Authorizer = authorizer
 
+	for role, scale := range scaleOps {
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: Post-deploy, detect new VMs and wait for them to be ready.  The kernel for all of this is in updatePlusOne, it just needs to be factored out.  I think it's much better for us just to await readiness on new VMs.

Comment (jim-minter) [reply cmd/createorupdate/update.go]: *Then*, if we were just doing a scale, return early.

File: pkg/validate/validate.go [hunk 10]
```diff
@@ -68,7 +69,29 @@ func validateContainerService(c *api.OpenShiftManagedCluster, externalOnly bool)
...
+		if !ok {
+			continue
+		}
+		if old.Count != new.Count {
+			isScaleUpdate = true
+		}
+		oldCs.Properties.AgentPoolProfiles[i].Count = new.Count
+	}
+
+	if isScaleUpdate && oldCs.Properties.OrchestratorProfile.OrchestratorVersion != cs.Properties.OrchestratorProfile.OrchestratorVersion {
```
Comment (jim-minter) [inline pkg/validate/validate.go]: This is possibly premature.

Comment (pweil-) [reply pkg/validate/validate.go]: I'm sure someone will try to do it.  :-) Seems good if we want to limit it from the beginning

Comment (kargakis) [reply pkg/validate/validate.go]: @jim-minter I thought we wanted to start with as little update surface as possible :)

File: cmd/createorupdate/update_test.go [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+		name      string
+		cs, oldCs *api.OpenShiftManagedCluster
+		exp       map[api.AgentPoolProfileRole]scaleOp
+	}{
+		{
+			name: "no change",
+			cs: &api.OpenShiftManagedCluster{
+				Properties: &api.Properties{
+					OrchestratorProfile: &api.OrchestratorProfile{},
+				},
```
Comment (pweil-) [inline cmd/createorupdate/update_test.go]: this doesn't actually test the counts not changing

File: pkg/validate/validate.go (hunk 10 above)
Comment (jim-minter) [reply pkg/validate/validate.go]: yes; at the moment we're not even allowing OrchestratorVersion to change.  I'd rather we start allowing it to change at the point where we have some confidence that it works.

File: cmd/createorupdate/update.go [hunk 12]
```diff
@@ -43,6 +70,68 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
...
+		if newCounts[old.Name] > old.Count {
+			scaleOps[old.Role] = scaleOp{up: true}
+			hasScaleOps = true
+			log.Debug("scale up")
+		} else if newCounts[old.Name] < old.Count {
+			scaleOps[old.Role] = scaleOp{down: true}
+			hasScaleOps = true
+			log.Debug("scale down")
+		}
+		oldCs.Properties.AgentPoolProfiles[i].Count = newCounts[old.Name]
```
Comment (pweil-) [inline cmd/createorupdate/update.go]: why are we mutating `oldCs` here?

Comment (pweil-) [reply cmd/createorupdate/update.go]: oh, for the deepEqual.  This seems bad, anything after this that uses the pointer isn't going to have accurate info

Comment (kargakis) [reply cmd/createorupdate/update.go]: Agreed, we are in need of deep copies to do these kind of operations safely. For now, this should be fine since once we figure out a difference in the counts, we don't care about old anymore.

File: cmd/createorupdate/update.go (hunk 7 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Depends on whether we want to allow scale+upgrade in a single update.

File: cmd/createorupdate/update.go (hunk 9 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Does it matter if we just check readiness for all VMs now? Tidying up to check new VMs only can be a follow-up?

File: pkg/validate/validate.go (hunk 3 above)
Comment (kargakis) [reply pkg/validate/validate.go]: It's much uglier though because we need to figure out a difference in counts in two different places: 1. validation (here) and 2. before apply the ARM template. So mutating counts in validation is a no-no. For now, I am copying the counts back after running deep equal. Eventually we should have a real deep copy for OpenShiftManagedCluster.

File: cmd/createorupdate/update.go (hunk 7 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Btw how do we know that we are draining the correct VM that will get deleted during the ARM update? Is it always the highest index that gets removed first?

File: cmd/createorupdate/update_test.go (hunk 11 above)
Comment (kargakis) [reply cmd/createorupdate/update_test.go]: fixed

File: cmd/createorupdate/update.go (hunk 8 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: done

File: cmd/createorupdate/update.go (hunk 9 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: Fixed the readiness check. Still outstanding is whether we scale+upgrade in one go.

File: pkg/validate/validate.go (hunk 10 above)
Comment (kargakis) [reply pkg/validate/validate.go]: > yes; at the moment we're not even allowing OrchestratorVersion to change. I'd rather we start allowing it to change at the point where we have some confidence that it works.

Seems like we need to discuss whether this is something we want to have for PP (don't we want image updates?). FWIW, when I started testing upgrades a couple of weeks ago, I was able to update from one image to another w/o any problem.

File: cmd/createorupdate/update.go (hunk 7 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: This is not an issue after all, we just need to drain+delete the right number of VMs.

File: pkg/validate/validate.go [hunk 13]
```diff
@@ -68,9 +69,40 @@ func validateContainerService(c *api.OpenShiftManagedCluster, externalOnly bool)
 }
 
 func validateUpdateContainerService(cs, oldCs *api.OpenShiftManagedCluster, externalOnly bool) (errs []error) {
-	if !reflect.DeepEqual(cs, oldCs) {
+	newAgents := make(map[string]*api.AgentPoolProfile)
+	for _, newAgent := range cs.Properties.AgentPoolProfiles {
+		newAgents[newAgent.Name] = newAgent
```
[the remark this answers is not in the document]
Comment (jim-minter) [reply pkg/validate/validate.go]: WIP at #316

File: cmd/createorupdate/update.go [hunk 14]
```diff
@@ -43,6 +106,49 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
 	return nil
 }
 
+type scaleOp struct {
+	up   bool
+	down bool
+	// diff is the count of VMs that are going
+	// to be scaled up or down.
+	diff int
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: pretty sure you can't do this - it's not idempotent

File: cmd/createorupdate/update.go [hunk 15]
```diff
@@ -43,6 +106,49 @@ func update(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.Plugin)
 	return nil
 }
 
+type scaleOp struct {
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: I don't really like this structure at all

File: cmd/createorupdate/update.go [hunk 16]
```diff
@@ -11,18 +12,80 @@ import (
...
 
+	// Need to determine whether the current update is a scale operation before
+	// applying the ARM template. For scale up, we need to figure out which are
+	// the new VMs that were created and wait for them to turn ready. For scale
+	// down, we need to drain the correct VMs before applying the ARM template
+	// scales them down.
+	scaleOps := isScaleOp(cs, oldCs)
+	vmsBefore := map[string]struct{}{}
+	for role, scale := range scaleOps {
+		vms, err := listVMs(ctx, cs, vmc, role)
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: should we not just always do this for all the roles?

File: pkg/api/types.go [hunk 17]
```diff
@@ -119,6 +119,27 @@ type AgentPoolProfile struct {
 	Role         AgentPoolProfileRole `json:"role,omitempty"`
 }
 
+// TODO: Get rid of this in favor of generated deep copies
+func (a *AgentPoolProfile) DeepCopy() *AgentPoolProfile {
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/api/types.go]: agreed

File: cmd/createorupdate/update.go (hunk 14 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: I can get rid of this but in order to make scale indempotent we always need to check for the health of all VMs, right?

Comment (kargakis) [reply cmd/createorupdate/update.go]: Can you walk me through a scenario where this is not idempotent?

File: cmd/createorupdate/update.go (hunk 16 above)
Comment (jim-minter) [reply cmd/createorupdate/update.go]: the key action here is draining VMs that deploy() is about to cause to be deleted.  A potential case could be after a previously infra/compute upgrade failed where there are currently N+1 VMs all serving workload.
I think it'd be highly unexpected for this code to be doing anything with master VMs - you could just iterate infra and compute roles for now.

Comment (jim-minter) [reply cmd/createorupdate/update.go]: I think this is the case where storing the diff is invalid - should be storing the number of VMs that is expected post-deploy.

File: cmd/createorupdate/update.go (hunk 14 above)
Comment (jim-minter) [reply cmd/createorupdate/update.go]: I don't think you need to store the delta here in order to be able solely to health check newly created VMs.  You do need to list the VMs pre- and post- deploy though but that code is elsewhere.

File: cmd/createorupdate/update.go (hunk 15 above)
Comment (jim-minter) [reply cmd/createorupdate/update.go]: I don't think it adds anything - if you have it, you need additional unit tests, but if you don't have it, the unit tests aren't really lacking.
Clearly, you need a unit test around the function which detects if a diff between cs and oldCs is scale only.
Otherwise, if you robustly drain all excess VMs before deploy() and health check all new VMs post deploy(), then there is no diff-related code to unit test.  It's more robust and you're just in the desired-state->actual state reconciliation world.

Comment (kargakis) [reply cmd/createorupdate/update.go]: >  Clearly, you need a unit test around the function which detects if a diff between cs and oldCs is scale only.

If you have a function that detects you have a scale operation, you still need some extra logic to figure out the direction you are scaling. The point of `isScaleOp` is exactly that, to detect whether the current update is a scale operation and furthermore whether it's up or down. In order to make it unit-testable it has to be its own function. When `isScaleOp` returns, `update` needs to know what scale operation it is about to perform on which scale set, hence `isScaleOp` returns a `map[api.AgentPoolProfileRole]scaleOp` which should always be of size 1 but I don't do any sort of validation about it atm. So I am not sure how to better model this and make as much logic unit-testable as possible, other than use `scaleOp`. Probably holding `count` vs `diff` is more robust but that does not diminish our need for `scaleOp`, does it?

File: cmd/createorupdate/update.go (hunk 14 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: FWIW, this is only used by scale down where we don't need to list post-deploy (deploy is actually a no-op). I went ahead and updated it to the desired count of the scale set.

File: pkg/validate/validate.go [hunk 18]
```diff
@@ -68,9 +69,40 @@ func validateContainerService(c *api.OpenShiftManagedCluster, externalOnly bool)
...
+	}
+	oldVersion := oldCs.Properties.OrchestratorProfile.OrchestratorVersion
+	oldCs.Properties.OrchestratorProfile.OrchestratorVersion = cs.Properties.OrchestratorProfile.OrchestratorVersion
+
+	if !reflect.DeepEqual(cs.Properties, oldCs.Properties) {
 		errs = append(errs, fmt.Errorf("invalid change %s", deep.Equal(cs, oldCs)))
 	}
+	// Restore whatever we mutated because some of it (eg. scales) are needed
+	// down the road.
+	// TODO: Deep-copy OpenShiftManagedCluster
```
[the remark this answers is not in the document]
Comment (kargakis) [reply pkg/validate/validate.go]: done

File: cmd/createorupdate/update.go (hunk 16 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: By "diff", do you mean the field in `scaleOp` which I just removed in favor of `scaleOp.count` or the diff between new and old VMs that we do now in case of a scale up to figure out which are the new VMs we should wait for? In case of the latter, I am not sure we have any use for the number of desired VMs post-deploy even in the case we are coming from a failed upgrade with a N+1 VMs currently running. `deploy` during the current scale up should either be a no-op (because we scaled up to the actual VMs) or we should scale up past the surge VM. In both cases, the diff between new and old VMs should still work as expected (we should wait for the VMs created during this scale up).

**Comment by kargakis** [comment]
I've pushed a commit that removes `scaleOp.diff` in favor of `scaleOp.count` (we need to know how many VMs to remove during a scale down). This should be ready for another review. In a follow up I want to create interfaces for the VM, SSC clients, and our Plugin can use a fake implementation, to make the whole `update` fn unit-testable.

**Comment by kargakis** [comment]
/hold

@jim-minter ptal in the last commit

File: cmd/createorupdate/update.go [hunk 19]
```diff
@@ -11,18 +12,70 @@ import (
...
+	isScaleOp := isScaleUpdate(cs, oldCs)
+	vmsBefore := map[string]struct{}{}
+	if isScaleOp {
+		for _, agent := range cs.Properties.AgentPoolProfiles {
+			vms, err := listVMs(ctx, cs, vmc, agent.Role)
+			if err != nil {
+				return err
+			}
+
+			for _, vm := range vms[agent.Count:] {
```
Comment (kargakis) [inline cmd/createorupdate/update.go]: Getting a slice bount out of range with this
```
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
main.update(0x12bd3a0, 0xc42039e540, 0xc420383620, 0xc420452f00, 0x12c7600, 0xc42000e628, 0xc4205fe000, 0x2bad3, 0x54f6c, 0x0, ...)
	/home/kargakis/go/src/github.com/openshift/openshift-azure/cmd/createorupdate/update.go:40 +0x1065
main.createOrUpdate(0x12bd3a0, 0xc42039e540, 0xc4203830e0, 0xc4203225a0, 0xc42037bef0, 0xc420390810, 0x12bd3a0)
	/home/kargakis/go/src/github.com/openshift/openshift-azure/cmd/createorupdate/createorupdate.go:104 +0x481
main.main()
	/home/kargakis/go/src/github.com/openshift/openshift-azure/cmd/createorupdate/createorupdate.go:206 +0x7f5
exit status 2
```

Comment (kargakis) [reply cmd/createorupdate/update.go]: Happens when I want to scale up

File: cmd/createorupdate/update.go [hunk 20]
```diff
@@ -2,27 +2,84 @@ package main
...
+	}
+
+	// Apply the ARM template
+	if err := deploy(ctx, cs, p, azuredeploy); err != nil {
+		return err
+	}
+
+	if isScaleOp {
+		for _, agent := range cs.Properties.AgentPoolProfiles {
+			vms, err := listVMs(ctx, cs, vmc, agent.Role)
```
Comment (kargakis) [inline cmd/createorupdate/update.go]: Another thing we avoided with `isScaleOp` was listing after deploy when we are scaling down.

**Comment by kargakis** [comment]
/hold

Needs proper squashing
